automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/notebooks/WorkspaceNotebooksPage.scala

```diff
@@ -0,0 +1,207 @@
+package org.broadinstitute.dsde.firecloud.page.workspaces.notebooks
+
+import com.typesafe.scalalogging.LazyLogging
+import org.broadinstitute.dsde.firecloud.FireCloudView
+import org.broadinstitute.dsde.firecloud.component.Component._
+import org.broadinstitute.dsde.firecloud.component._
+import org.broadinstitute.dsde.firecloud.page.PageUtil
+import org.broadinstitute.dsde.firecloud.page.workspaces.WorkspacePage
+import org.broadinstitute.dsde.workbench.config.Config
+import org.broadinstitute.dsde.workbench.model.google.GcsPath
+import org.broadinstitute.dsde.workbench.service.test.WebBrowserUtil
+import org.openqa.selenium.WebDriver
+import org.scalatest.selenium.{Page, WebBrowser}
+
+class WorkspaceNotebooksPage(namespace: String, name: String)(implicit webDriver: WebDriver)
+  extends WorkspacePage(namespace, name) with Page with PageUtil[WorkspaceNotebooksPage] with LazyLogging {
+
+  override def awaitReady(): Unit = clustersTable.awaitReady()
+
+  override val url: String = s"${Config.FireCloud.baseUrl}#workspaces/$namespace/$name/notebooks"
+
+  private val clustersTable = Table("spark-clusters-table")
+  private val clustersTableTestId = testId("spark-clusters-table")
+  private val clusterErrorMessage = testId("notebooks-error")
+  private val sparkClustersHeader = testId("spark-clusters-title")
+  private val openCreateClusterModalButton: Button = Button("create-modal-button")
+  private val noClustersMessage = "There are no clusters to display."
+
+  //  withNewCluster - open modal, create cluster, wait for it to appear/running state, open delete modal, delete cluster, wait for delete
+  def withNewCluster(clusterName: String,
+                     masterMachineType: String = "n1-standard-4",
+                     masterDiskSize: Int = 500,
+                     workers: Int = 0,
+                     workerMachineType: String = "n1-standard-4",
+                     workerDiskSize: Int = 500,
+                     workerLocalSSDs: Int = 0,
+                     preemptibleWorkers: Int = 0,
+                     extensionURI: Option[GcsPath] = None,
+                     customScriptURI: Option[GcsPath] = None,
+                     labels: Map[String, String] = Map())(testCode: (WorkspaceNotebooksPage) => Any) = {
+
+    openCreateClusterModal.createCluster(clusterName, masterMachineType, masterDiskSize, workers, workerMachineType, workerDiskSize, workerLocalSSDs, preemptibleWorkers, extensionURI, customScriptURI, labels)
+    await text clusterName
+    assert(getClusterStatus(clusterName) == "Creating")
+    logger.info("Creating dataproc cluster " + clusterName)
+    waitUntilClusterIsRunning(clusterName)
+    logger.info("Created dataproc cluster " + clusterName)
+
+    testCode
+
+    openDeleteClusterModal(clusterName).deleteCluster
```

Comment:
It would be nice to do some defensive cleanup here. Right now if `testCode` fails, then the cluster would never be deleted. Could do something like Leo's `withNewCluster`:

https://github.com/DataBiosphere/leonardo/blob/develop/automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoTestUtils.scala#L222

Reply:
I feel like this code that checks all the different conditions should be in the Spec test class. I think that it could be useful to change this withNewCluster method to use the APIs to create the cluster and in that case it should use a register cleanUp to call the delete API. Then maybe this withNewCluster call can be used to verify we can create clusters with all the different machine types in the Spec class.

Reply:
Good practice to have asserts/verifications on test class-level.